### ISSUE TITLE
Version 0.4

### ISSUE THREAD

**Issue by dladams** [LSSTDESC/desc-prod#4]
For version 4, I have promised to add support for LSST pipeline processing.

We make use of the existing [gen3_worflow](https://github.com/LSSTDESC/gen3_workflow) and [ctrl_bps](https://github.com/lsst/ctrl_bps) packages.

For now the supporting code will be put in this package (desc-prod) to keep things simple.

**Comment by dladams** [comment]
I did resurrected my old butler test script raw.py (in ~dladams/desc/butler) that does queries for raw data datasets and print their IDs. A couple observations:
* When I set up from cvmfs, it can be slow and the first time I import ExposureF and Butler on a node, it takes 30-60 sec. Only 0.5 sec after that.
* I some times see very long delays (many minutes) when constructing the Butler object. I suspect a DB problem.
* My script dos not run for weekly builds w_2023_16 due to problems with ExposureF. 

Once things were working, I saw 4.25M raw datasets in the DC2 data:
<pre>
repo = '/global/cfs/cdirs/lsst/production/gen3/DC2/Run2.2i/repo'
col = 'LSSTCam-imSim/raw/all'
butler = Butler(repo, collections=[col])
</pre>
Here are IDs for a few:
<pre>
{band: 'z', instrument: 'LSSTCam-imSim', detector: 93, physical_filter: 'z_sim_1.4', exposure: 1212554}
{band: 'i', instrument: 'LSSTCam-imSim', detector: 12, physical_filter: 'i_sim_1.4', exposure: 1217146}
{band: 'i', instrument: 'LSSTCam-imSim', detector: 144, physical_filter: 'i_sim_1.4', exposure: 1231245}
{band: 'r', instrument: 'LSSTCam-imSim', detector: 71, physical_filter: 'r_sim_1.4', exposure: 1173504}
</pre>

**Comment by dladams** [comment]
I added am trying to build a dockerfile for gen3_workflow following the [prescription](https://github.com/LSSTDESC/gen3_workflow/blob/master/doc/Package_requirements.rst) in that package but the solve is slow (30 min and counting).

**Comment by dladams** [comment]
I finally got the docker file to build by switching from ndctools 7.3.4 to 7.3.5. This is in dockerfile-02. I tried many conda incantations which made things worse. Switching the solver gives a small improvement: 670 s instea of 710 s for the cond install. This is in dockerfile-03.

The docker builds scripts are in docker/gen3workflow.

Changes are in 0.4.1.

**Comment by dladams** [comment]
Also in 0.4.1 is dockerfile04 which adds the last two lines in setup file installed at /opt/lsst/software/stack/setup.sh:
<pre>
source /opt/lsst/software/stack/loadLSST.bash
setup lsst_apps
export PS1='lsst:w_2023_15> '
export LSSTVERSION='w_2023_15> '
export OMP_NUM_THREADS=1
</pre>
I created image dladams/gen3workflow-w_2023_15:02, pushed it to docker and installed in perlmutter shifter. I am in the process of doing this for the current weekly in image dladams/gen3workflow-w_2023_18:04.
Thsese dockerfiles were constructed following the instructions in the [gen3_workflow documentation](https://github.com/LSSTDESC/gen3_workflow/blob/master/doc/Package_requirements.rst)

**Comment by dladams** [comment]
For 0.4.2, I add scripts that can be used to setup the gen3 env using either cvmfs or shifter images. These are in desc-prod/apps/gen3setup and include:
<pre>
g3wf-setup-cvmfs - Sets up the gen3 env using a cvmfs release in the current shell
g3wf-run-cvmfs - Starts a new gen3 cvmfs shell using this setup
g3wf-run-shifter - Starts a new shell with the gen3 env from a shifter image
g3wf-setup-cvmfs - Python script that queries DC2 data for images (test program)
runapp-gewfquery - Application interface for jobtype g3wfquery which runs that script
</pre>
The latter is intended as test of the scripts to set up the gen3 env and enable users to begin to examine the DC2 data.

The shifter script uses the images described above.
The cvmfs setup uses LSST cvmfs installation plus a condor env with gen3 extensions again following instructions in the [gen3_workflow documentation](https://github.com/LSSTDESC/gen3_workflow/blob/master/doc/Package_requirements.rst).

**Comment by dladams** [comment]
I decided to move the gen3 scripts to a dedicated package [desc-gen3-prod](https://github.com/LSSTDESC/desc-gen3-prod). Progress there is being tracked in [desc-gen3-prod issue 1](https://github.com/LSSTDESC/desc-gen3-prod/issues/1).

**Comment by dladams** [comment]
For 0.4.3, the gen3 docker stuff is removed after copying to desc-gen3-prod.

**Comment by dladams** [comment]
I updated the packger readme including adding a link to desc-gen3-prod.

**Comment by dladams** [comment]
I updated the readme, apps.md and the server help page. The 2nd has condiderably more information about how jobs progress through the system and how to write applications.

**Comment by dladams** [comment]
For version 0.0.6, I bring over some my old gen3 wrapper code and begin to create application g3wfpipe to run pipeline jobs.

After many hours and fixes, I am apparently able to start a pipeline job. But it is out of control: 199 processes with 1-2% of the total CPU. I have to kill many processes to quiet this.

**Comment by dladams** [comment]
After struggling a bit trying to understand why my jobs would not start properly yesterday, I checked the MOTD and found perlmutter was down (although I was able log in and issue start job commands). I used the opportunity to fix the perlmutter status button. Code changes (not needed for button to work) are in 0.4.4. See #5.

**Comment by dladams** [comment]
Typing the full config (jobtype, config, howfig) for each job has proved to be cumbersome. To reduce this effort, the following are introduced for version 0.4.5:

1. The new job form on the primary web page (jobtype, config, howfig) is now prefilled with the values stored in the user data. When a session starts, these are set using the highest numbered job for the user.
2. After a job is successfully created, these starting values are set to the values for that job.
3. The menu for each job (click on job ID) now has an entry "Copy job". If this is selected the starting values are set from that job.

**Comment by dladams** [comment]
For version 0.4.7, the command descprod-output-dir is added to return a directory for descprod output for the current user on the current machine. Default is $PSCRATCH.

**Comment by dladams** [comment]
For version 0.4.8, change descprod-add-job to use subdirectory jobs of descprod-output-dir as the base for run directories instead of the local directory. Users can recover the old behavior by setting DESCPROD_OUTPUT_DIR=. before adding a job.

**Comment by dladams** [comment]
in 0.4.9, "-p" was added in descprod-output-dir.

**Comment by dladams** [comment]
in 0.4.10, All progress messages are logged with time stamps in progress.log in job directories.

**Comment by dladams** [comment]
For 0.4.11, service crashes when user logs in:
<pre>
 10.42.3.44 - - [31/Jul/2023 18:36:36] "GET /home HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/home/descprod/conda/lib/python3.10/site-packages/flask/app.py", line 2551, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/descprod/conda/lib/python3.10/site-packages/flask/app.py", line 2531, in wsgi_app
    response = self.handle_exception(e)
  File "/home/descprod/conda/lib/python3.10/site-packages/flask/app.py", line 2528, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/descprod/conda/lib/python3.10/site-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/descprod/conda/lib/python3.10/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/descprod/conda/lib/python3.10/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/home/descprod/dev/desc-prod/server/app.py", line 288, in home
    jtab = JobTable(udat.descname)
  File "/home/descprod/conda/lib/python3.10/site-packages/descprod/job_table.py", line 13, in __init__
    self.refresh()
  File "/home/descprod/conda/lib/python3.10/site-packages/descprod/job_table.py", line 16, in refresh
    self.jobs = JobData.get_jobs_from_db(self.descname)
  File "/home/descprod/conda/lib/python3.10/site-packages/descprod/jobs.py", line 145, in get_jobs_from_db
    cur = cls.db_query_where(f"descname='{descname}'", cols='id')
  File "/home/descprod/conda/lib/python3.10/site-packages/descprod/jobs.py", line 366, in db_query_where
    haveit = cls.db_table(tnam)
  File "/home/descprod/conda/lib/python3.10/site-packages/descprod/jobs.py", line 241, in db_table
    raise Exception(f"{myname}: Unable to connect to DB {dbnam}")
Exception: JobData.db_table: Unable to connect to DB descprod
10.42.3.44 - - [31/Jul/2023 18:36:36] "GET /home?__debugger__=yes&cmd=resource&f=style.css HTTP/1.1" 304 -
10.42.3.44 - - [31/Jul/2023 18:36:36] "GET /home?__debugger__=yes&cmd=resource&f=debugger.js HTTP/1.1" 304 -
10.42.3.44 - - [31/Jul/2023 18:36:36] "GET /home?__debugger__=yes&cmd=resource&f=console.png HTTP/1.1" 304 -
</pre>
I see in the log and when I try by hand that the mysql service is not starting. I opened a [ticket](https://nersc.servicenowservices.com/sp/?sys_id=548e07311b98f15059b0dc6ce54bcbc7&view=sp&id=ticket&table=incident) to address that issue.

In addition, I would like have the service start with a useful error message rather than crash when the mysql service is absent or times out. The server code is modified so this message is displayed in the web page:
<pre>
ERROR: Job DB query failed.
</pre>
instead of raising an exception.

**Comment by dladams** [comment]
I found a way to fix the mysql problem at NERSC. I modify server/start so that option mysql:noaio will configure mysqld with
<pre>
innodb_use_native_aio = 0
</pre>
before starting mysql. The NERSC production service is configured to use this option instead of "mysql".
Code change is in 0.4.12.

**Comment by dladams** [comment]
For 0.4.14, I added follow opts to descprod-start job:
* -f follows the job log file
* -p follows the progress log

**Comment by dladams** [comment]
Version 0.4.16 add options to follow logs when starting jobs with _descprod-start-job_.

**Comment by dladams** [comment]
Version 0.4.17 is used to resolve #10.

**Comment by dladams** [comment]
Version 0.4.18 has developments for laptop server (perlmutter is down today):
* In docker start (docker/descprod/start), add option -descprod to set user to descprod.
* In docker start, make version 10 the default.
* In server start (server/start) map opt update to update-desc-prod instead of that, wfmon and parsl.

But mysqld would not start. Log showed a timeout error when trying to upgrade DB from 8.0.32 to 8.0.34.
I started the service once with --upgrade=MINIMAL and the problem went away. I added an equivalent option in the web server start script (minup) in case we need to do this again.

This problem presumably arose because my latest image descprod:10 has the new version of mysql, release in August I believe.

I also found I have to run the container as descprod instead of root for the job table to be visible to the web service.

Laptop service is noiw working with image descprod:10

**Comment by dladams** [comment]
For 0.4.19, add GUI buttons to disable refresh of focus and to toggle though a set of refresh periods including off.

**Comment by dladams** [comment]
For 0.4.20 fix #11

**Comment by dladams** [comment]
0.4.21, 0.4.22 and 0.4.23 were used to fix the problem of sessions appearing in the wrong browser (#7).
Tags v0.4.20 and v0.4.23 were made for those releases.

**Comment by dladams** [comment]
Version 0.4.24 (tag v0.4.24) is another fix described at #7.

**Comment by dladams** [comment]
And 0.4.25 has more for #7.

**Comment by dladams** [comment]
0.4.26 adds SSL validation for 2024 SSL certificate. This can be dropped but I keep it as an example for next year.

**Comment by dladams** [comment]
0.4.27 is the same as 0.4.26. I made a mod trying to get SSL working but put things back when I found the real problem (chain was not installed).

**Comment by dladams** [comment]
For 0.4.28, add job field _archive_ to be nonzero if a job has been archived. By default the user home page will show jobs that are not archived.

**Comment by dladams** [comment]
For 0.0.49, the job wrapper is updated (now version 19) to protect against interrupts. See #14.
